lib/logstash/outputs/s3.rb

```diff
@@ -73,285 +61,353 @@
+    configure_upload_workers
 
-  # if you lose connection with s3, bad control implementation.
-  if ( @s3 == nil)
-    aws_s3_config
+    @codec.on_event do |event, encoded_event|
+      handle_event(encoded_event)
+    end
   end
 
-  # find and use the bucket
-  bucket = @s3.buckets[@bucket]
-
-  @logger.debug "S3: ready to write "+file_basename+" in bucket "+@bucket+", Fire in the hole!"
-
-  # prepare for write the file
-  object = bucket.objects[file_basename]
-  object.write(:file => file_data, :acl => @canned_acl)
-
-  @logger.debug "S3: has written "+file_basename+" in bucket "+@bucket + " with canned ACL \"" + @canned_acl + "\""
-
- end
+  public
+  def configure_upload_workers
+    @logger.debug("S3: Configure upload workers")
 
- # this method is used for create new path for name the file
- def getFinalPath
+    @upload_workers = @upload_workers_count.times.map do |worker_id|
+      Stud::Task.new do
+        LogStash::Util::set_thread_name("<S3 upload worker #{worker_id}")
 
-   @pass_time = Time.now
-   return @temp_directory+"ls.s3."+Socket.gethostname+"."+(@pass_time).strftime("%Y-%m-%dT%H.%M")
+        while true do
+          @logger.debug("S3: upload worker is waiting for a new file to upload.", :worker_id => worker_id)
 
- end
-
- # This method is used for restore the previous crash of logstash or to prepare the files to send in bucket.
- # Take two parameter: flag and name. Flag indicate if you want to restore or not, name is the name of file
- def upFile(flag, name)
+          upload_worker
+        end
+      end
+    end
+  end
 
-   Dir[@temp_directory+name].each do |file|
-     name_file = File.basename(file)
+  private
+  def upload_worker
+    file = @upload_queue.deq
+
+    case file
+      when LogStash::ShutdownEvent
+        @logger.debug("S3: upload worker is shutting down gracefuly")
+        @upload_queue.enq(LogStash::ShutdownEvent)
+      else
+        @logger.debug("S3: upload working is uploading a new file", :filename => File.basename(file))
+        move_file_to_bucket(file)
+    end
+  end
 
-     if (flag == true)
-      @logger.warn "S3: have found temporary file: "+name_file+", something has crashed before... Prepare for upload in bucket!"
-     end
+  public
+  def next_page
+    @page_counter += 1
+  end
 
-     if (!File.zero?(file))
-       write_on_bucket(file, name_file)
+  def reset_page_counter
+    @page_counter = 0
+  end
 
-       if (flag == true)
-          @logger.debug "S3: file: "+name_file+" restored on bucket "+@bucket
-       else
-          @logger.debug "S3: file: "+name_file+" was put on bucket "+@bucket
-       end
-     end
+  # Use the same method that Amazon use to check
+  # permission on the user bucket by creating a small file
+  public
+  def test_s3_write
+    @logger.debug("S3: Creating a test file on S3")
 
-     File.delete (file)
+    test_filename = File.join(@temporary_directory, "logstash-programmatic-access-test-object")
```

Comment:
After thinking, this would make sure the check works if something goes wrong.
I'll add it.